[Dong Cheng-Yu] iP

officialchengyud opened the pull request:
### HarvardBot

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

HarvardBot frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

- download it from [here](https://nus-cs2103-ay2324s2.github.io/website/schedule/week4/project.html).
- double-click it.
- add your tasks.
- let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

officialchengyud committed:
Level 0. Rename, Greet, Exit

officialchengyud committed:
Improve the skeletal version of Duke so that it echos commands entered by the user, and exits when the user types the command bye.

officialchengyud committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested.

officialchengyud committed:
Add the ability to mark tasks as done. Optionally, add the ability to change the status back to not done.

officialchengyud committed:
Add support for tracking three types of tasks:
1. ToDos: tasks without any date/time attached to it
2. Deadlines: tasks that need to be done before a specific date/time
3. Events: tasks that start at a specific date/time and ends at a specific date/time

officialchengyud committed:
Use the input/output redirection technique to semi-automate the testing of Duke.

officialchengyud committed:
Teach the chatbot to deal with errors such as incorrect inputs entered by the user.

officialchengyud committed:
Add support for deleting tasks from the list.

officialchengyud committed:
Add save feature

This allows retention of tasks after ending program.

The tasks are saved in a text file located ./data/Harvard.txt

It is a simple and easy way to store data without a database.

officialchengyud committed:
Add support for dates

Allows for more scalability of the chat bot.

Stores deadline dates as java.time.LocalDate

This allows for more options when displaying dates.

officialchengyud committed:
Refactor code to make it OOP

This follows better standard Java practices

Extracting essential features and splitting into classes

This is a more preferable paradigm

officialchengyud committed:
Divide classes into packages

Makes the project much more organised and better adheres to OOP practice

Split classes into packages

officialchengyud committed:
Merge branch 'add-gradle-support'

officialchengyud committed:
Add JUnit tests

Need to test behaviour of code to ensure.

Add JUnit tests for different classes.

JUnit follows standard practice for testing code behaviour.

officialchengyud committed:
Package the App as a JAR file

officialchengyud committed:
Add JavaDocs to methods and classes

Allow other coders to understand the code

Added JavaDoc to all classes and public methods

JavaDoc is the documentation standard

officialchengyud committed:
Tweak code to follow coding standard

Allows other people to collaborate on this code more easily

Follow SE-EDU coding standard practices

This is the standard decided for this project

officialchengyud committed:
Add find feature

Makes it easier for users to find tasks

Add "find" command which filters task names

This allows for efficient searching

officialchengyud committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/harvard/Parser.java
#	src/main/java/harvard/Storage.java
#	src/main/java/harvard/TaskList.java

officialchengyud committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/harvard/Parser.java
#	src/main/java/harvard/TaskList.java

ararchch commented on src/main/java/harvard/Storage.java:
```diff
@@ -0,0 +1,82 @@
+     *
+     * @param tasks The TaskList containing tasks to be stored.
+     */
+    public void store(TaskList tasks) {
```
Could making the name of the method more descriptive improve readability? For example `storeTasksInFile`

ararchch commented on src/main/java/harvard/tasks/Task.java:
```diff
@@ -0,0 +1,36 @@
+        return this.description;
+    }
+
+    public String saveString() {
```
Could this method's function be made more clear with a different name?

ararchch commented on src/main/java/harvard/TaskList.java:
```diff
@@ -0,0 +1,148 @@
+    /**
+     * The list that stores tasks.
+     */
+    private List<Task> taskList = new ArrayList<>();
```
Would it be more readable to use a plural term like `tasks` to represent the list of tasks?

ararchch commented on src/test/java/harvard/TaskListTest.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Test
+    public void taskList_printTaskFound_success() {
+        TaskList tL = new TaskList();
```
Could the name tL here be longer to make it more readable and easy to follow?

ararchch commented on src/main/java/harvard/Ui.java:
```diff
@@ -0,0 +1,131 @@
+    /**
+     * Displays a welcome message when the application starts.
+     */
+    public void showWelcome() {
```
I like how you have used nouns for things and verbs for actions everywhere!

ararchch reviewed (commented):
Great job! LGTM, just some small nits regarding code quality and naming of variables.

yorklim commented on data/harvard.txt:
Should this be committed? Can consider adding this to .gitignore

yorklim commented on src/main/java/Harvard.java:
Looks good to me just some minor nits, can consider arranging the imports in lexical order

yorklim commented on src/main/java/harvard/Ui.java:
```diff
@@ -0,0 +1,131 @@
+     * Displays a welcome message when the application starts.
+     */
+    public void showWelcome() {
+        String initial = "____________________________________________________________\n"
```
Can consider using a string line break constant instead of doing "_______" which can help to make code cleaner

yorklim commented on src/main/java/harvard/Storage.java:
```diff
@@ -0,0 +1,82 @@
+     *
+     * @return The BufferedReader containing task information.
+     */
+    public BufferedReader load() {
```
I like that you created the file storage location right at the start if the file is not detected instead of waiting till you save the save

yorklim commented on src/main/java/harvard/Storage.java:
```diff
@@ -0,0 +1,82 @@
+     * @param fP The file path for storing tasks.
+     */
+    private String filePath;
+    public Storage(String fP) {
```
Can consider naming the function parameters fP clearer

yorklim commented on src/main/java/harvard/Parser.java:
Any reason why some of the classes you used this.variable but some you just do variable?

yorklim reviewed (commented):
Overall LGTM just maybe some minor nits!

officialchengyud committed:
Add CheckStyle

Ensure the consistency of styling in the application

Add gradle checkstyle with custom settings

This helps developers to ensure entire project has consistent styling

officialchengyud committed:
Add JavaFX technology

Implement GUI for better appearance

Use JavaFX technology in the project

JavaFX is ideal for creating GUI in this chatbot project

Pughal77 commented on src/main/java/harvard/Parser.java:
```diff
@@ -0,0 +1,127 @@
+    public String parse(String commandLine) throws HarvardException {
+        String command = commandLine.split(" ")[0];
+
+        if (!command.equals("list") && !command.equals("todo")
```
Would use of switch-case be better here? Any reason you went for if-else rather than switch-case?

Pughal77 commented on src/main/java/harvard/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+     * @return The string representation of the task.
+     * @throws HarvardException If the index is out of bounds.
+     */
+    public String printString(int index) throws HarvardException {
```
Perhaps you want to return the task itself instead of a string representation?

Pughal77 commented on src/main/java/harvard/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+     * @param searchString the string to search for within task descriptions
+     * @return a TaskList containing tasks whose descriptions contain the search string
+     */
+    public TaskList find(String searchString) {
```
Nice! You returned a TaskList so that the String representation is followed.

Pughal77 commented on src/main/java/harvard/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+     */
+    public String printString(int index) throws HarvardException {
+        if (index > this.taskList.size() - 1) {
+            throw new HarvardException("Sorry, this task could not be found.");
```
Perhaps you would like to return it as a response using the ui?

Pughal77 reviewed (commented):
LGTM. Just a few nit picks

Pughal77 commented on src/main/java/harvard/Parser.java:
```diff
@@ -0,0 +1,127 @@
+     * @param commandLine The user command to be parsed.
+     * @throws HarvardException If an error occurs during parsing or task manipulation.
+     */
+    public String parse(String commandLine) throws HarvardException {
```
Perhaps userInput or commandInput is a clearer naming?

Pughal77 commented on src/main/java/harvard/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+     *
+     * @param buffReader The BufferedReader containing task information.
+     */
+    public void populateTaskList(BufferedReader buffReader) {
```
perhaps the naming of beffreader could be changed to lineInText to improve readability?

Pughal77 commented on src/main/java/harvard/tasks/Todo.java:
```diff
@@ -0,0 +1,50 @@
+     * @return a string representing the Todo object in a format suitable for saving
+     */
+    @Override
+    public String saveString() {
```
Perhaps the naming for this method could be saveToTextFile? Instead of saying what the method does perhaps its better to say what the method is for.
```suggestion
    public String saveToTextFile() {
```

Pughal77 reviewed (commented):
LGTM. A few nit picks.

officialchengyud committed:
Use Assertions

Documents important assumptions at various points in code

Use java assert within various parts of code

Java assert is a good feature to use

officialchengyud committed:
Examine and refactor code

Ensure code quality

Look through code and check for any code violations

To ensure the overall quality of code for the entire project

officialchengyud committed:
Add better search

Easier for users to search for tasks

Checks for partial matches in task name rather for exact match

Good way to find tasks more easily

officialchengyud committed:
Merge pull request #2 from officialchengyud/branch-A-Assertions

Use Assertions

officialchengyud committed:
Merge pull request #3 from officialchengyud/branch-A-CodeQuality

Examine and refactor code

officialchengyud committed:
Merge pull request #4 from officialchengyud/branch-C-BetterSearch

Add better search

officialchengyud committed:
Refine GUI and update profile photos

Improve the user experience

Add padding to text and changed some colours

Looks better and more refined for the user

officialchengyud committed:
Merge pull request #5 from officialchengyud/branch-A-BetterGui

Refine GUI and update profile photos

officialchengyud committed:
Add brief user guide

Help users understand how the program works

Add commands into the user guide markdown

Easier to understand the commands and features of the chatbot

officialchengyud committed:
Update README.md

officialchengyud committed:
Add error checking

Ensure user entered commands are checked

Check the dates and format of command

Alert users of any issues with the entered command

officialchengyud committed:
Merge pull request #6 from officialchengyud/branch-A-MoreErrorHandling

Add error checking
